qutebrowser/completion/models/sqlmodel.py

```diff
@@ -0,0 +1,273 @@
+        _categories: The category tables.
+    """
+
+    def __init__(self, column_widths=(30, 70, 0), columns_to_filter=None,
+                 delete_cur_item=None, parent=None):
+        super().__init__(parent)
+        self.columns_to_filter = columns_to_filter or [0]
+        self.column_widths = column_widths
+        self._categories = []
+        self.srcmodel = self  # TODO: dummy for compat with old API
+        self.pattern = ''
+        self.delete_cur_item = delete_cur_item
+
+    def _cat_from_idx(self, index):
+        """Return the category pointed to by the given index.
+
+        Args:
+            idx: A QModelIndex
+        Returns:
+            A _SqlCompletionCategory if the index points at one, else None
+        """
+        # items hold an index to the parent category in their internalPointer
+        # categories have an empty internalPointer
+        if index.isValid() and not index.internalPointer():
+            return self._categories[index.row()]
+
+    def new_category(self, name, select='*', where=None, sort_by=None,
+                     sort_order=None, limit=None):
+        """Create a new completion category and add it to this model.
+
+        Args:
+            name: Name of category, and the table in the database.
+            select: A custom result column expression for the select statement.
+            where: An optional clause to filter out some rows.
+            sort_by: The name of the field to sort by, or None for no sorting.
+            sort_order: Sorting order, if sort_by is non-None.
+            limit: Maximum row count to return on a query.
+
+        Return: A new CompletionCategory.
+        """
+        cat = _SqlCompletionCategory(name, parent=self, sort_by=sort_by,
+                                     sort_order=sort_order, limit=limit,
+                                     select=select, where=where,
+                                     columns_to_filter=self.columns_to_filter)
+        self._categories.append(cat)
+
+    def data(self, index, role=Qt.DisplayRole):
+        """Return the item data for index.
+
+        Override QAbstractItemModel::data.
+
+        Args:
+            index: The QModelIndex to get item flags for.
+
+        Return: The item data, or None on an invalid index.
+        """
+        if not index.isValid() or role != Qt.DisplayRole:
+            return
+        if not index.parent().isValid():
+            if index.column() == 0:
+                return self._categories[index.row()].tablename
+        else:
+            table = self._categories[index.parent().row()]
+            idx = table.index(index.row(), index.column())
+            return table.data(idx)
+
+    def flags(self, index):
+        """Return the item flags for index.
+
+        Override QAbstractItemModel::flags.
+
+        Return: The item flags, or Qt.NoItemFlags on error.
+        """
+        if not index.isValid():
+            return
```

Comment:
`return Qt.NoItemFlags`? From what it looks like, I messed this up in `CompletionModel` too. I think it should be fine (as both end up as `0` when casted), but it'd be more correct.